pkg/services/ngalert/state/state.go

```diff
@@ -129,83 +180,77 @@ func NewEvaluationValues(m map[string]eval.NumberValueCapture) map[string]*float
+	if state.State == eval.Normal {
+		logger.Debug("Keeping state", "state", state.State)
+	} else {
 		logger.Debug("Changing state", "previous_state", state.State, "next_state", eval.Normal)
-		state.State = eval.Normal
-		state.StartsAt = result.EvaluatedAt
-		state.EndsAt = result.EvaluatedAt
+		// Normal states have the same start and end timestamps
+		state.Normal(result.EvaluatedAt, result.EvaluatedAt)
 	}
 }
 
 func resultAlerting(state *State, rule *models.AlertRule, result eval.Result, logger log.Logger) {
-	state.Error = result.Error
-
 	switch state.State {
 	case eval.Alerting:
-		// If the previous state is Alerting then update the expiration time
-		state.setEndsAt(rule, result)
+		logger.Debug("Keeping state", "state", state.State)
+		state.Maintain(rule.IntervalSeconds, result.EvaluatedAt)
 	case eval.Pending:
 		// If the previous state is Pending then check if the For duration has been observed
 		if result.EvaluatedAt.Sub(state.StartsAt) >= rule.For {
 			logger.Debug("Changing state", "previous_state", state.State, "next_state", eval.Alerting)
-			state.State = eval.Alerting
-			state.StartsAt = result.EvaluatedAt
-			state.setEndsAt(rule, result)
+			state.Alerting(result.EvaluatedAt, nextEndsTime(rule.IntervalSeconds, result.EvaluatedAt))
 		}
 	default:
 		if rule.For > 0 {
 			// If the alert rule has a For duration that should be observed then the state should be set to Pending
 			logger.Debug("Changing state", "previous_state", state.State, "next_state", eval.Pending)
-			state.State = eval.Pending
+			state.Pending(result.EvaluatedAt, nextEndsTime(rule.IntervalSeconds, result.EvaluatedAt))
 		} else {
 			logger.Debug("Changing state", "previous_state", state.State, "next_state", eval.Alerting)
-			state.State = eval.Alerting
+			state.Alerting(result.EvaluatedAt, nextEndsTime(rule.IntervalSeconds, result.EvaluatedAt))
 		}
-		state.StartsAt = result.EvaluatedAt
-		state.setEndsAt(rule, result)
 	}
 }
-
 func resultError(state *State, rule *models.AlertRule, result eval.Result, logger log.Logger) {
 	switch rule.ExecErrState {
 	case models.AlertingErrState:
+		logger.Debug("Execution error state is Alerting", "handler", "resultAlerting", "previous_handler", "resultError")
 		resultAlerting(state, rule, result, logger)
-	case models.ErrorErrState:
+		// This is a special case where Alerting and Pending should also have an error and reason
 		state.Error = result.Error
-		if result.Error != nil {
-			// If the evaluation failed because a query returned an error then add the Ref ID and
-			// Datasource UID as labels
-			var queryError expr.QueryError
-			if errors.As(state.Error, &queryError) {
-				for _, next := range rule.Data {
-					if next.RefID == queryError.RefID {
-						state.Labels["ref_id"] = next.RefID
-						state.Labels["datasource_uid"] = next.DatasourceUID
-						break
-					}
-				}
-				state.Annotations["Error"] = queryError.Error()
-			}
-		}
-
+		state.StateReason = "error"
```

Comment:
Please create a constant in models (there is already one)

Reply:
Also, I think we can add error as StateReason. The same for OK mapping for the Error state

Reply:
Done 🙂